### PR TITLE
Feat/pull request generator

### PR DESCRIPTION
**Pull request by shmurata**
The new generator proposed at https://github.com/argoproj-labs/applicationset/issues/210 has been implemented as a Pull Request Generator.
Although the Pull Request Generator is similar to the SCM Provider Generator in some aspects, it seems that various changes are necessary to incorporate it into the SCM Provider Generator, so I have implemented it as a separate generator.
In the future, it may be possible to merge them.

I've implemented it to add other providers as well, but this PR only supports Github.
I can't implement other providers because I don't use them.

The requeue interval is polled in this PR, but I'd like to add webhook support to this generator after https://github.com/argoproj-labs/applicationset/pull/341 is merged.

### CODE REVIEW

File: test/e2e/applicationset/applicationset_test.go [hunk 1]
```diff
@@ -508,3 +508,72 @@ func TestCustomApplicationFinalizers(t *testing.T) {
...
+		When().Create(v1alpha1.ApplicationSet{ObjectMeta: metav1.ObjectMeta{
+		Name: "simple-pull-request-generator",
+	},
+		Spec: v1alpha1.ApplicationSetSpec{
+			Template: v1alpha1.ApplicationSetTemplate{
+				ApplicationSetTemplateMeta: v1alpha1.ApplicationSetTemplateMeta{Name: "guestbook-{{ number }}"},
+				Spec: argov1alpha1.ApplicationSpec{
+					Project: "default",
+					Source: argov1alpha1.ApplicationSource{
+						RepoURL:        "git@github.com:shmurata/argocd-example-apps.git",
```
Comment (jgwest) [inline test/e2e/applicationset/applicationset_test.go:551]: I've created a new test org for ApplicationSet repo to uses for test like this:
https://github.com/applicationset-test-org/argocd-example-apps

I've copied over the contents of your repository (the branches, and the PR) to this new org. Are you able to update the test to use this new org? Take a look and let me know if I missed something :sweat_smile:

File: docs/Generators.md [hunk 2]
```diff
@@ -12,5 +12,6 @@ As of this writing there are six generators:
 - [Matrix generator](Generators-Matrix.md): The Matrix generator may be used to combine the generated parameters of two separate generators.
 - [SCM Provider generator](Generators-SCM-Provider.md): The SCM Provider generator uses the API of an SCM provider (eg GitHub) to automatically discover repositories within an organization.
 - [Cluster Decision Resource generator](Generators-Cluster-Decision-Resource.md): The Cluster Decision Resource generator is used to interface with Kubernetes custom resources that use custom resource-specific logic to decide which set of Argo CD clusters to deploy to.
+- [Pull Request generator](Generators-Pull-Request.md): The Pull Request generator uses the API of an SCMaaS provider (eg GitHub) to automatically discover opening pull requests within an repository.
```
Comment (jgwest) [inline docs/Generators.md:15]: Small tweak to this line:
`- [Pull Request generator](Generators-Pull-Request.md): The Pull Request generator uses the API of an SCMaaS provider (eg GitHub) to automatically discover open pull requests within an repository.`

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great @shmurata! A couple of tweaks requested... and it looks like you need to add the new generator doc file to `mkdocs.yml` as well.

**Comment by shmurata** [comment]
@jgwest  Thanks for the review.
I would like to respond to your comments.
If there is anything else you would like to see fixed or recommend, it would be appreciated if you could comment on it.

File: test/e2e/applicationset/applicationset_test.go (hunk 1 above)
Comment (shmurata) [reply test/e2e/applicationset/applicationset_test.go:551]: Thank you for preparing the repository.
I would like to ask for one additional setting.
In this test, we are also checking the label filter. Can you please add a label to the pull request that you prepared?

File: docs/Generators-Pull-Request.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
+# Pull Request Generator
+
+The Pull Request generator uses the API of an SCMaaS provider (eg GitHub) to automatically discover opening pull requests within an repository. This fits well with the style of building a test environment when you create a pull request.
```
[the remark this answers is not in the document]
Comment (shmurata) [reply docs/Generators-Pull-Request.md]: Fixed in 62c44b9.

File: docs/Generators.md (hunk 2 above)
Comment (shmurata) [reply docs/Generators.md]: Fixed in 62c44b9.

**Comment by shmurata** [comment]
Added to mkdocs.yml in 6e81ce5.

File: test/e2e/applicationset/applicationset_test.go (hunk 1 above)
Comment (jgwest) [reply test/e2e/applicationset/applicationset_test.go:551]: @shmurata Done.

Comment (shmurata) [reply test/e2e/applicationset/applicationset_test.go]: Thanks. Fix in 6bbe516.

**Review by jgwest — APPROVED** [review]
Approved, thanks @shmurata , and thanks for your patience :sweat_smile: .

**Comment by enys** [comment]
Thanks for this !

**Pull request merged**